File: app/ios/NeurOne/Models/ResearchSuggestion.swift
import Foundation

// Research suggestion portal model — CLAUDE.md §6.3.
// Three functions: (1) patient research agenda, (2) pre-identified subject pool, (3) crowdfunding catalyst.
// Server calls are stubs; backend infrastructure is out of scope for T1 (see ISA scope exclusion §2.3).

// MARK: - Suggestion

struct ResearchSuggestion: Codable, Identifiable {
    var id: UUID
    var title: String
    var body: String
    var categories: [ResearchCategory]
    var submittedAt: Date
    var authorDeviceToken: String    // opaque per §5.3 — no user identity

    // Server-side aggregates (refreshed from API; optimistic local updates on vote/pledge)
    var voteCount: Int
    var participationIntentCount: Int
    var pledgedAmount: Int           // cumulative dollars pledged across all users

    // Per-user state (local — never transmitted as individual fields)
    var didVote: Bool
    var hasParticipationIntent: Bool
    var pledgeAmount: Int?           // nil = no pledge

    var fundingStatus: FundingStatus

    enum FundingStatus: String, Codable {
        case open       // collecting intent + pledges
        case active     // formal campaign live — escrow charged if target met
        case funded     // target met
        case closed     // campaign ended — refunded if target not met
    }
}

// MARK: - Submitted idea draft (local, pre-server)

struct ResearchSuggestionDraft {
    var title: String = ""
    var body: String = ""
    var categories: Set<ResearchCategory> = []
    var hasParticipationIntent: Bool = false
    var pledgeAmount: Int = 0        // 0 = no pledge

    var isValid: Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !categories.isEmpty
    }
}

// MARK: - Pledge tier helpers

enum PledgeTier: Int, CaseIterable {
    case none = 0
    case small = 10
    case medium = 25
    case large = 50
    case max = 100

    var label: String {
        switch self {
        case .none:   return "No pledge"
        case .small:  return "$10"
        case .medium: return "$25"
        case .large:  return "$50"
        case .max:    return "$100+"
        }
    }
}

File: app/android/core/src/main/kotlin/life/neurone/core/research/ResearchSuggestion.kt
package life.neurone.core.research

import life.neurone.core.models.ResearchCategory
import kotlinx.serialization.Serializable
import java.util.UUID

// Port of iOS ResearchSuggestion.swift (CLAUDE.md §6.3). Three functions: (1) patient
// research agenda, (2) pre-identified subject pool, (3) crowdfunding catalyst. Server sync is
// a stub for T1 — this is local state only. `authorDeviceToken` is an opaque per-device token
// (no user identity). Exact timestamps are avoided: day granularity only, consistent with the
// privacy posture elsewhere.

@Serializable
enum class FundingStatus { OPEN, ACTIVE, FUNDED, CLOSED }

@Serializable
data class ResearchSuggestion(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val body: String,
    val categories: List<ResearchCategory>,
    val submittedAtDay: String,        // "yyyy-MM-dd"
    val authorDeviceToken: String,     // opaque — no user identity
    // Server-side aggregates (optimistic local updates on vote/pledge).
    val voteCount: Int = 0,
    val participationIntentCount: Int = 0,
    val pledgedAmount: Int = 0,        // cumulative dollars pledged across all users
    // Per-user state (local — never transmitted as individual fields).
    val didVote: Boolean = false,
    val hasParticipationIntent: Boolean = false,
    val pledgeAmount: Int? = null,     // null = no pledge
    val fundingStatus: FundingStatus = FundingStatus.OPEN,
)

// Local, pre-server draft.
data class ResearchSuggestionDraft(
    val title: String = "",
    val body: String = "",
    val categories: Set<ResearchCategory> = emptySet(),
    val hasParticipationIntent: Boolean = false,
    val pledgeAmount: Int = 0,         // 0 = no pledge
) {
    val isValid: Boolean
        get() = title.isNotBlank() && body.isNotBlank() && categories.isNotEmpty()
}

enum class PledgeTier(val amount: Int, val label: String) {
    NONE(0, "No pledge"),
    SMALL(10, "\$10"),
    MEDIUM(25, "\$25"),
    LARGE(50, "\$50"),
    MAX(100, "\$100+"),
}
